service/resource/deployment/resource.go

```diff
@@ -187,7 +190,85 @@ func (r *Resource) GetDeleteState(ctx context.Context, obj, currentState, desire
 }
 
 func (r *Resource) GetUpdateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, interface{}, interface{}, error) {
-	return nil, nil, nil, nil
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+	currentDeployments, err := toDeployments(currentState)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+	desiredDeployments, err := toDeployments(desiredState)
+	if err != nil {
+		return nil, nil, nil, microerror.Mask(err)
+	}
+
+	var deploymentsToCreate interface{}
+	{
+		deploymentsToCreate, err = r.GetCreateState(ctx, obj, currentState, desiredState)
+		if err != nil {
+			return nil, nil, nil, microerror.Mask(err)
+		}
+	}
+
+	var deploymentsToDelete []*v1beta1.Deployment
+	{
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "finding out which deployments have to be deleted")
+
+		for _, currentDeployment := range currentDeployments {
+			if !containsDeployment(desiredDeployments, currentDeployment) {
+				deploymentsToDelete = append(deploymentsToDelete, currentDeployment)
+			}
+		}
+
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", fmt.Sprintf("found %d deployments that have to be deleted", len(deploymentsToDelete)))
+	}
+
+	var deploymentsToUpdate []*v1beta1.Deployment
+	if updateallowedcontext.IsUpdateAllowed(ctx) {
+		r.logger.Log("cluster", key.ClusterID(customObject), "debug", "finding out which deployments have to be updated")
+
+		// Check if config maps of deployments changed. In case they did, add the
+		// deployments to the list of deployments intended to be updated.
+		m, ok := messagecontext.FromContext(ctx)
+		if ok {
+			for _, name := range m.ConfigMapNames {
+				desiredDeployment, err := getDeploymentByConfigMapName(desiredDeployments, name)
+				if err != nil {
+					return nil, nil, nil, microerror.Mask(err)
+				}
+				deploymentsToUpdate = append(deploymentsToUpdate, desiredDeployment)
+			}
+		}
+
+		// Check if deployments changed. In case they did, add the deployments to
+		// the list of deployments intended to be updated, but only in case they are
+		// not already being tracked.
+		for _, currentDeployment := range currentDeployments {
```

Comment:
Why not looping over currentDeployments only once? We already loop above when getting the `deploymentsToDelete`

Reply:
I do not like mixing functionality. As the comments describe we select deployments we want to update based on different criteria. For me the code is more easily understandable to do one thing at a time and another thing afterwards. You are right that we could theoretically do multiple things at once. Anyway, for me personally the code is more structured that way. Like, there are different parameters the function has to return, so I compute them separately, because the parameters are separated.

Reply:
I see, maybe the function has more than one responsibility, could it be split in three (smaller! ;)) functions, one for each parameter?

Reply:
I would keep it as it is here. I agree that this should be split a bit in case it is getting more complicated. We could also think about the interface of the resource and if we could maybe improve on this level to prevent such problems.